File: book/snippets/python/src/event/kind.py
from typing import cast
from nostr_sdk import Kind, KindEnum, EventBuilder, Keys, Metadata


def kind():
    print()
    keys = Keys.generate()
    print("Kind:")

    # ANCHOR: kind-int
    print("  Kind from integer:")
    kind = Kind(1)
    print(f"     - Kind 1: {kind.as_enum()}")
    kind = Kind(0)
    print(f"     - Kind 0: {kind.as_enum()}")
    kind = Kind(3)
    print(f"     - Kind 3: {kind.as_enum()}")
    # ANCHOR_END: kind-int

    print()
    # ANCHOR: kind-enum
    print("  Kind from enum:")
    kind = Kind.from_enum(cast(KindEnum, KindEnum.TEXT_NOTE()))
    print(f"     - Kind TEXT_NOTE: {kind.as_u16()}")
    kind = Kind.from_enum(cast(KindEnum, KindEnum.METADATA()))
    print(f"     - Kind METADATA: {kind.as_u16()}")
    kind = Kind.from_enum(cast(KindEnum, KindEnum.CONTACT_LIST()))
    print(f"     - Kind CONTRACT_LIST: {kind.as_u16()}")
    # ANCHOR_END: kind-enum

    print()
    # ANCHOR: kind-methods
    print("  Kind methods EventBuilder:")
    event  = EventBuilder.text_note("This is a note").sign_with_keys(keys)
    print(f"     - Kind text_note(): {event.kind().as_u16()} - {event.kind().as_enum()}")
    event  = EventBuilder.metadata(Metadata()).sign_with_keys(keys)
    print(f"     - Kind metadata(): {event.kind().as_u16()} - {event.kind().as_enum()}")
    event  = EventBuilder.contact_list([]).sign_with_keys(keys)
    print(f"     - Kind contact_list(): {event.kind().as_u16()} - {event.kind().as_enum()}")
    # ANCHOR_END: kind-methods

    print()
    # ANCHOR: kind-representations
    kind = Kind(1337)
    print(f"Custom Event Kind: {kind.as_u16()} - {kind.as_enum()}")
    # ANCHOR_END: kind-representations

    print()
    # ANCHOR: kind-tests
    print("  Kind Logical Tests:")
    kind = Kind(30001)
    print(f"     - Is {kind.as_u16()} addressable?: {kind.is_addressable()}")
    kind = Kind(20001)
    print(f"     - Is {kind.as_u16()} ephemeral?: {kind.is_ephemeral()}")
    kind = Kind(5001)
    print(f"     - Is {kind.as_u16()} job request?: {kind.is_job_request()}")
    kind = Kind(6001)
    print(f"     - Is {kind.as_u16()} job result?: {kind.is_job_result()}")
    kind = Kind(1)
    print(f"     - Is {kind.as_u16()} regular?: {kind.is_regular()}")
    kind = Kind(10001)
    print(f"     - Is {kind.as_u16()} relay replaceable?: {kind.is_replaceable()}")
    # ANCHOR_END: kind-tests

File: book/snippets/js/src/event/kind.ts
import { Kind, Keys, EventBuilder, Metadata, loadWasmSync } from "@rust-nostr/nostr-sdk"

function kind() {
    // Load WASM
    loadWasmSync();

    // Generate random keys
    let keys = Keys.generate();

    console.log();
    console.log("Kind:");

    // ANCHOR: kind-int
    console.log("  Kind by number:");
    let kind = new Kind(1);
    console.log(`     - Kind 1: ${kind.toString()}`);
    kind = new Kind(0);
    console.log(`     - Kind 0: ${kind.toString()}`);
    kind = new Kind(3);
    console.log(`     - Kind 3: ${kind.toString()}`);
    // ANCHOR_END: kind-int

    console.log();
    // ANCHOR: kind-methods
    console.log("  Kind methods EventBuilder:");
    let event = EventBuilder.textNote("This is a note").signWithKeys(keys);
    console.log(`     - Kind textNote(): ${event.kind.asU16()}`);
    event = EventBuilder.metadata(new Metadata()).signWithKeys(keys);
    console.log(`     - Kind metadata(): ${event.kind.asU16()}`);
    event = EventBuilder.contactList([]).signWithKeys(keys);
    console.log(`     - Kind contactList(): ${event.kind.asU16()}`);
    // ANCHOR_END: kind-methods

    console.log();
    // ANCHOR: kind-tests
    console.log("  Kind Logical Tests:");
    kind = new Kind(30001);
    console.log(`     - Is ${kind.toString()} addressable?: ${kind.isAddressable()}`);
    kind = new Kind(20001);
    console.log(`     - Is ${kind.toString()} ephemeral?: ${kind.isEphemeral()}`);
    kind = new Kind(5001);
    console.log(`     - Is ${kind.toString()} job request?: ${kind.isJobRequest()}`);
    kind = new Kind(6001);
    console.log(`     - Is ${kind.toString()} job result?: ${kind.isJobResult()}`);
    kind = new Kind(1);
    console.log(`     - Is ${kind.toString()} regular?: ${kind.isRegular()}`);
    kind = new Kind(10001);
    console.log(`     - Is ${kind.toString()} replaceable?: ${kind.isReplaceable()}`);
    // ANCHOR_END: kind-tests
}

kind();
